src/Palette.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/inclusive-design/adaptive-palette/blob/main/LICENSE
+ */
+
+import { render } from "preact";
+import { html } from "htm/preact";
+import { PaletteCell } from "./PaletteCell";
+import "./Palette.scss";
+
+/**
+ * Given a palette defined in a json structure, compute the number of rows
+ * and columns in that palette.
+ *
+ * @param {Object} paletteDefinition - An object that lists the positions,
+ *                 heights and widths of the cells in the palette.
+ * @return {Object} - The row and column counts: `{ numRows: ..., numColumns: ...}`.
+ */
+function countRowsColumns (paletteDefinition) {
+  let rowCount = 0;
+  let colCount = 0;
+  let rightColumn = 0;
+  let bottomRow = 0;
+  const cellIds = Object.keys(paletteDefinition.cells);
+  cellIds.forEach((id) => {
+    const cellOptions = paletteDefinition.cells[id].options;
+    rightColumn = cellOptions.columnStart + cellOptions.columnSpan;
+    if (rightColumn > colCount) {
+      colCount = rightColumn;
+    }
+    bottomRow = cellOptions.rowStart + cellOptions.rowSpan;
+    if (bottomRow > rowCount) {
+      rowCount = bottomRow;
+    }
+  });
+  return { numRows: rowCount, numColumns: colCount };
+}
+
+export function Palette (props) {
+
+  const paletteDefinition = props.json;
+  const rowsCols = countRowsColumns(paletteDefinition);
+  const cellIds = Object.keys(paletteDefinition.cells);
+
+  // Loop to create an array of renderings for each cell
+  const theCells = [];
+  cellIds.forEach((id) => {
+    const aCell = paletteDefinition.cells[id];
+    const cellOptions = aCell.options;
+    const paletteCell = html`
+      <${PaletteCell} id="${id}" labelText="${cellOptions.label}"
```

Comment:
The `cellOptions` can be passed as one single prop so that when this line can be shared for instantiating various components of various cell types:

`<${PaletteCell} id="${id}" options=${cellOptions} />`

Reply:
Done.

I've also incorporated your latest `bmw_palette.json` with the correct layout of the BMW code keys.  I noticed that the `type` of key is not within the options block.  I have no opinion if it should be or not, but I'm passing it along like so:

```
<${PaletteCell} id="${id}" options=${cellOptions} type=${aCell.type} />
```

At the moment, `PaletteCell` does nothing with the type, but I presume it will be used when there is a component registry that maps type strings to components.

Reply:
Replying to myself:  I'm at the wrong level.  If there is a registry, the type will be used by the `Palette` component, something like:

```
<${aCell.type} id="${id}" options=${cellOptions} />
```
There is probably no reason to pass the `type` to the `PaletteCell` component.  I'll take it out at some point.